sabnzbd/downloader.py

```diff
@@ -428,7 +428,7 @@ def decode(self, article, raw_data):
                     sabnzbd.Assembler.queue.qsize(),
                 )
                 logged = True
-            time.sleep(self.sleep_time)
+            time.sleep(self.sleep_time + 0.001)
```

Comment:
I would like to suggest to keep the hardcoded values here, unless there's a real good reason to make it flexible?

Reply:
I don't really know, but my hypothesis was that on a fast connection and a fast CPU a short sleep is desirable, while on a slower CPU which uses more time on decoding and thread switching it is better with a higher value. 0.05 seems too long to me as it limits the speed to 320 KB/s (16KB buffer / 0.05) per connection. This may make the transfer rate choppy and slower than necessary. I don't mind hard coding it, but I would suggest have a lower value than 0.05. How about 0.005? Do you know how much time it takes in all to decode one article or clear for instance 10% of the decoder buffer on various systems? That might be a good starting point.

Reply:
Testing this on a raspberry pi and my super slow old Synology NAS I actually found it's rarely in this loop for too long. Maybe a few times per second. The slow CPU in general will just prevent things from going too fast. 
So I'm quite confident that the hard-coded value is a decent one :)

Reply:
I don't think I'm explaining this well, or maybe I'm just wrong. Anyway, I think the problem will be with fast CPUs and fast connections, not the slow ones. The NewsWrapper recv_chunk sets the buffer to 16KB for SSL. When it is full, I assume the server will quickly have to stop sending data, and the speed will slow down.

On a fast connection this takes a very short time. At 1Gbps, even with 50 connections you would fill all the 16KB buffers in about 0.006 seconds (total buffer = 50 connections * 16384 bytes * 8  = 6.25 Mbit. 6.25/1000 = 0.00625). If the CPU is just a little bit too slow for this speed then sleeping 0.05 seconds means it's waiting much longer than it needs to to keep up.

Reply:
Luckily the decoding is super fast and has its own internal queue, so on those systems it never hits this code. I can confirm with my i7 on a gigabit connection.

Reply:
Oh wait I see you referenced the buffer in newswrapper for receiving the chucks. Actually that's not effected here, since this code only checks for the decoder and assembler queues. The chunk buffer is only dependent on the main while loop and the select in there.

Reply:
The decode function which does the sleep is called from the downloader main while loop, so when you sleep in decode you pause the main loop for the same amount of time. All the packages that are received after the buffers have filled will be dropped, and the sender will assume it's sending too fast.

It's no problem if the CPU is fast enough, but it is when it's not. If it can decode 900 Mbps then it will start sleeping in decode. However, when you wait as much as 0.05s before clearing of the receive buffers then that's the equivalent of limiting the bandwidth to about 125 Mbps with 50 connections.

On my CPU it seems that one thread can decode one 800 KB article in about 0.008 seconds (total processing time in the DecoderWorker loop, not just the decode). That's about the maximum time it will take before there is a free spot in the queue. If it was almost done with the previous article it will take less time. That means it could frequently get back to checking the receive buffer in a much shorter time than 0.05s.

Reply:
Yes, that is if you assume every decode input will trigger a 0.05 delay, but that's not the case. When we wait 0.05 it will have decoded maybe 10 articles, so the sleep won't be called for the next 10 articles.
This is just a theoretical discussion, all I wanted to say is that the sleep here and in the speed limiter have a different impact than the one in the main loop because their purpose is different. So that's why I don't want to use the same constant.

Reply:
How about a compromise?
time.sleep(self.sleep_time * 10 + 0.001)

That defaults so 0.011s, which won't trigger the loops a lot more than now. It's much higher than my current one but gives some leeway in adjusting it both up and down for end users. That way those who experience slow downs can check if this is the problem.

From my own testing a with a 4MB/s transfer rate using 1 connection, a 0.05s sleep can start slowing it down if it triggers more often than 1 per second.